### DISCUSSION TITLE
aspnet-core-authentication-schemes/

### DISCUSSION THREAD

**Discussion by giscus** [matteocontrini/blog#1] [Comments]
# aspnet-core-authentication-schemes/

How to define custom authentication schemes in ASP.NET Core 6.0, and why they’re not enough to actually enforce authentication for your Web API.

https://blog.contrini.it/aspnet-core-authentication-schemes/

**Comment by Yerkon** [comment]
Nice description👍

**Comment by Yerkon** [comment]
One case that doesn't properly work is when we don't specify default authentication scheme in case when we have several authentication scheme. 
We have currently two auth. schemes, and I wanted to one of them work as default with just [Authorize] attribute. And  added: 
```
 services.AddAuthentication()
                .AddScheme<SchemeFirstOptions, SchemeFirstHandler>(
                    'First', options => {}
                )
                .AddScheme<SchemeSecondOptions, SchemeSecondHandler>(
                    'Second', options => {}
                )

builder.Services.AddAuthorization(options =>
{
    options.DefaultPolicy = new AuthorizationPolicyBuilder()
        .AddAuthenticationSchemes("First")
        .RequireAuthenticatedUser()
        .Build();
});
```
So, when: 
    [Authorize(AuthenticationSchemes = 'Second')]
The 'Second' custom auth. handler run, after that 'First' auth handler running. Not what we expect.

In case: 
 [Authorize(AuthenticationSchemes = 'Second', Policy='Abc')]
The 'Second' custom auth. handler run and done. (No 'First' scheme run). Which we expect.

So, without specifying policy default scheme is always running

**Reply by matteocontrini** [reply]
That's interesting. Although maybe not expected, it makes sense if you think about it. I'll make sure to mention this behavior in the blog post, thanks for your comment!

**Comment by Harwex213** [comment]
Very helpful. Thanks a lot

**Comment by marlonfsolis** [comment]
Where is the implementation of the TokenAuthenticationSchemeOptions? How I do implement it?

**Reply by matteocontrini** [reply]
Hey, the authentication scheme options is a simple class that you'd need to create and that must derive from `AuthenticationSchemeOptions`. If you don't require any options, you can just leave it empty.

Otherwise, it could be something like:

```csharp
public class SessionTokenAuthenticationSchemeOptions : AuthenticationSchemeOptions
{
    public bool MyOption { get; set; }
}
```

And you would then set that option when adding authentication:

```csharp
builder.Services.AddAuthentication()
    .AddScheme<SessionTokenAuthenticationSchemeOptions, SessionTokenAuthenticationSchemeHandler>(
        "SessionTokens",
        opts => {
            opts.MyOption = true; // <-- this line
        }
    );
```

Is it clearer now?

**Comment by MontyGvMC** [comment]
This article is really great and especially helpfull.

There is one thing that troubles me extending the functionality. I would like to use the options from configuration (appsettings/user secrets) rather than hardcoded.

I haven't found a way to configure or bind the SessionTokenAuthenticationSchemeOptions to the configuration. The hardcoded approach though works perfect.

**Reply by matteocontrini** [reply]
I see two possibilities:

1) Manually read settings and copy them to the options instance:

```csharp
builder.Services.AddAuthentication()
    .AddScheme<SessionTokenAuthenticationSchemeOptions, SessionTokenAuthenticationSchemeHandler>(
        "SessionTokens",
        opts => {
            opts.MyOption = builder.Configuration["Authentication:MyOption"]; // <-- this line
        }
    );
```

2) Bind the configuration to the options instance (this would also enable reload and stuff like that):

```csharp
builder.Services.AddAuthentication()
    .AddScheme<SessionTokenAuthenticationSchemeOptions, SessionTokenAuthenticationSchemeHandler>(
        "SessionTokens",
        opts => {
            builder.Configuration.GetSection("Authentication").Bind(opts); // <-- this line
        }
    );
```

I haven't run this code right now but it should work.

**Comment by stefanococco** [comment]
Hi, I have a question!
Is there a way to read session data from the HandleAuthenticateAsync method? I want to use sessions instead cookies to store logged user token.

Thanks :-)

**Reply by matteocontrini** [reply]
When you say sessions, you mean the built-in server-side sessions of ASP.NET Core?

I'm not very familiar with them but is there a reason why you can't just access them through the `HttpContext` (which is named `Context` if you inherit from `AuthenticationHandler`)?

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/app-state?view=aspnetcore-7.0

**Reply by stefanococco** [reply]
Oh yes, you're right! :-)

I accessed to session using Context, after adding builder.Services.AddSession() and app.UseSession() to program.cs.

**Comment by shinelev** [comment]
Thanks for great article!
It really helped me to put all puzzles into correct places and clear the whole picture!

**Comment by javed74** [comment]
Thanks for great and helpful article. Cleared lots of doubts and questions.

**Comment by hejiheji001** [comment]
Thanks for the great article. I have a question about how to pass the error message to the response body - searched a lot but noting related.

For example if the `return AuthenticateResult.Fail("Authentication failed");` is reached.

My current response body is like this, but I want to modify it - at least add the "Authentication failed" to the detail
`
{
    "type": "https://tools.ietf.org/html/rfc7235#section-3.1",
    "title": "Unauthorized",
    "status": 401,
    "detail": null,
    "instance": null,
    "extensions": {
        "traceId": "00-8e40804799f8ebc1bdcb0d3b76f4c2e9-a87cc49bc6dd5eec-00"
    }
}
`

**Reply by matteocontrini** [reply]
You can do that by overriding the `HandleChallengeAsync` method in the auth handler.

I'm using Minimal APIs so this is what I do inside that method:

```c#
protected override async Task HandleChallengeAsync(AuthenticationProperties properties)
{
    Response.StatusCode = 401;
    await Results.Json(new UnauthorizedError()).ExecuteAsync(Context);
}
```

where `UnauthorizedError` is a class of mine which gets automaticaly serialized to JSON using the minimal API serializer.

If you're using Web API (with controllers) I'm not sure if this works? If it doesn't, I think you have to resort to manually serialize the response and return it. Otherwise, you could probably override the global response status code handling, but that also depends on what you have in your `Program`.

I agree it should be easier though, it seems a basic thing to do.

**Comment by palapapa** [comment]
Absolutely amazing. The official documentation is terrible on this topic.

**Comment by therobotreader** [comment]
If you're like me and pulling your hair out because your claims aren't being registered, note that you have to `Add` them to the list of claims, not `Append` them.

As in, this works:
```
var claims = new List<Claim>();
claims.Add(new Claim('name', 'myname'));
```

and this doesn't:
```
var claims = new List<Claim>();
claims.Append(new Claim('name', 'myname'));
```

I have no idea why, but that's what worked for me.

**Reply by palapapa** [reply]
Because `Append` doesn't modify the original `IEnumerable`.
https://stackoverflow.com/questions/52355682/difference-between-a-lists-add-and-append-method

**Reply by therobotreader** [reply]
thanks!

**Comment by PierBover** [comment]
Great article. Thanks @matteocontrini !
